packages/compass-schema-validation/src/modules/validation.ts

```diff
@@ -144,280 +170,271 @@ export const checkValidator = (
-    error: action.validation.error || null,
-  };
-};
+  if (
+    isAction<ValidationSaveFailedAction>(
+      action,
+      ValidationActions.ValidationSaveFailed
+    )
+  ) {
+    return {
+      ...state,
+      error: action.error,
+    };
+  }
 
-/**
- * Set Error.
- */
-const setError = (
-  state: ValidationState,
-  action: ValidationSaveFailedAction
-): ValidationState => {
-  return {
-    ...state,
-    error: action.error || null,
-  };
-};
+  if (
+    isAction<EmptyValidationFetchedAction>(
+      action,
+      ValidationActions.EmptyValidationFetched
+    )
+  ) {
+    return {
+      ...state,
+      isChanged: false,
+      error: null,
+      syntaxError: null,
+      ...DEFAULT_VALIDATION,
+      validator: action.validationTemplate,
+    };
+  }
 
-/**
- * Change validation action.
- */
-const changeValidationAction = (
-  state: ValidationState,
-  action: ValidationActionChangedAction
-): ValidationState => {
-  const newState = {
-    ...state,
-    validationAction: action.validationAction,
-  };
+  if (
+    isAction<ValidationFetchErroredAction>(
+      action,
+      ValidationActions.ValidationFetchErrored
+    )
+  ) {
+    return {
+      ...state,
+      prevValidation: {
+        ...DEFAULT_VALIDATION,
+      },
+      isChanged: true,
+      error: action.error,
+      syntaxError: null,
+      ...DEFAULT_VALIDATION,
+    };
+  }
 
-  return {
-    ...newState,
-    isChanged: !isEqual(
-      pick(newState, ['validator', 'validationAction', 'validationLevel']),
-      state.prevValidation
-    ),
-  };
-};
+  if (
+    isAction<ValidationFetchedAction>(
+      action,
+      ValidationActions.ValidationFetched
+    )
+  ) {
+    // TODO(COMPASS-4989): EJSON??
+    const checkedValidator = checkValidator(
+      EJSON.stringify(action.validator, undefined, 2)
+    );
+    // TODO(COMPASS-4989): javascriptStringify??
+    const validator = javascriptStringify(
+      checkedValidator.validator,
+      null,
+      2
+    ) as string;
+
+    return {
+      ...state,
+      prevValidation: {
+        validator,
+        validationAction: action.validationAction,
+        validationLevel: action.validationLevel,
+      },
+      isChanged: false,
+      validator,
+      validationAction: action.validationAction,
+      validationLevel: action.validationLevel,
+      syntaxError: null,
+      error: null,
+    };
+  }
 
-/**
- * Change validation level.
- */
-const changeValidationLevel = (
-  state: ValidationState,
-  action: ValidationLevelChangedAction
-): ValidationState => {
-  const newState = {
-    ...state,
-    validationLevel: action.validationLevel,
-  };
+  if (
+    isAction<ValidationActionChangedAction>(
+      action,
+      ValidationActions.ValidationActionChanged
+    )
+  ) {
+    return {
+      ...state,
+      validationAction: action.validationAction,
+      isChanged: !isEqual(
+        pick(state, ['validator', 'validationAction', 'validationLevel']),
+        state.prevValidation
```

Comment:
For the initial change after you fetch `state.validationAction` is the same as `state.prevValidation.validationAction`, easy to repro if you start editing and change the action, I guess this should be comparing to the value from the action, not from the current state that's not updated yet at this point (this is also derived state, so I'd like to mention [this doc](https://redux.js.org/style-guide#keep-state-minimal-and-derive-additional-values) again)

Reply:
🤦 nice catch! Agreed that the derived `isChanged` would be better to calculate/memoize off of the state, it would introduce some more changes in this pr though. I'll take a look and see how much it would be. It would be nice to have this store a bit more minimal. `syntaxError` is another derived value that could be updated in a similar fashion.

Reply:
Yeah, no rush to immediately refactor this, but really good to keep in mind when thinking about redux stores, these types of issues are more or less specifically what keeping state minimal helps to address 🙂